content/courses/program-security/reinitialization-attacks.md

```diff
@@ -2,98 +2,109 @@
-actions. Creating an account involves invoking the `create_account` instruction
-handler on the System Program, which allocates space, rent in lamports, and
-assigns the program owner. Initialization sets the account data. These steps can
-be combined into a single transaction.
+Initialization refers to setting the data of a new account for the first time.
+When initializing a new account, you should implement a way to check if the
+account has already been initialized. Without an appropriate check, an existing
+account could be reinitialized and have existing data overwritten.
+
+Note that initializing an account and creating an account are two separate
+instructions. Creating an account requires invoking the `create_account`
+instruction on the System Program which specifies the space required for the
+account, the rent in lamports allocated to the account, and the program owner of
+the account. Initialization is an instruction that sets the data of a newly
+created account. Creating and initializing an account can be combined into a
+single transaction.
 
-### Missing Initialization Check
+#### Missing Initialization Check
 
-In the example below, there's no check on the `user` account. The `initialize`
-instruction handler sets the `authority` field on the `User` account type and
-serializes the data. Without checks, an attacker could reinitialize the account,
-overwriting the existing `authority`.
+In the example below, there are no checks on the `user` account. The
+`initialize` instruction deserializes the data of the `user` account as a `User`
+account type, sets the `authority` field, and serializes the updated account
+data to the `user` account.
+
+Without checks on the `user` account, the same account could be passed into the
+`initialize` instruction a second time by another party to overwrite the
+existing `authority` stored on the account data.
 
 ```rust
 use anchor_lang::prelude::*;
 
 declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
 
 #[program]
-pub mod initialization_insecure {
+pub mod initialization_insecure  {
     use super::*;
 
     pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
-        ctx.accounts.user.authority = ctx.accounts.authority.key();
+        let mut user = User::try_from_slice(&ctx.accounts.user.data.borrow()).unwrap();
+        user.authority = ctx.accounts.authority.key();
+        user.serialize(&mut *ctx.accounts.user.data.borrow_mut())?;
         Ok(())
     }
 }
 
 #[derive(Accounts)]
 pub struct Initialize<'info> {
+		#[account(mut)]
+    user: AccountInfo<'info>,
     #[account(mut)]
-    pub user: Account<'info, User>,
-    #[account(mut)]
-    pub authority: Signer<'info>,
-    pub system_program: Program<'info, System>,
+		authority: Signer<'info>,
 }
 
 #[account]
 #[derive(InitSpace)]
 pub struct User {
-    pub authority: Pubkey,
+    authority: Pubkey,
 }
 ```
 
-### Add is_initialized Check
+#### Add `is_initialized` check
```

Comment:
See https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#heading-styles. Sadly we have a technology limit here.